Implement URLPattern API's Constructor String Parser

jelee53 opened the pull request:
#### b98e269fd5a43b23d4e64683493a3960b46ae070
<pre>
Implement URLPattern API&apos;s Constructor String Parser
<a href="https://bugs.webkit.org/show_bug.cgi?id=284911">https://bugs.webkit.org/show_bug.cgi?id=284911</a>
<a href="https://rdar.apple.com/141717073">rdar://141717073</a>

Reviewed by Chris Dumez.

Spec: <a href="https://urlpattern.spec.whatwg.org/#constructor-string-parser">https://urlpattern.spec.whatwg.org/#constructor-string-parser</a>

* LayoutTests/fast/urlpattern/urlpattern-component-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern-compare.tentative.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern-compare.tentative.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern-compare.tentative.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern-compare.tentative.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern-compare.tentative.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern-compare.tentative.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern-compare.tentative.https.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern-compare.tentative.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.worker-expected.txt:
* Source/WebCore/Modules/url-pattern/URLPattern.cpp:
(WebCore::URLPattern::create):
(WebCore::URLPattern::compileAllComponents):
* Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
* Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp: Added.
(WebCore::URLPatternConstructorStringParser::URLPatternConstructorStringParser):
(WebCore::URLPatternConstructorStringParser::rewind):
(WebCore::URLPatternConstructorStringParser::getSafeToken const):
(WebCore::URLPatternConstructorStringParser::isNonSpecialPatternChararacter const):
(WebCore::URLPatternConstructorStringParser::isSearchPrefix const):
(WebCore::URLPatternConstructorStringParser::isAuthoritySlashesNext const):
(WebCore::URLPatternConstructorStringParser::makeComponentString const):
(WebCore::setInitComponentFromState):
(WebCore::URLPatternConstructorStringParser::computeProtocolMatchSpecialSchemeFlag):
(WebCore::URLPatternConstructorStringParser::changeState):
(WebCore::URLPatternConstructorStringParser::updateState):
(WebCore::URLPatternConstructorStringParser::performParse):
(WebCore::URLPatternConstructorStringParser::parse):
* Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.h: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/288189@main">https://commits.webkit.org/288189@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d7a3a09c4a681b868b3957c52cc2bfa5e3c7a3e3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82083 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1613 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36046 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86643 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33121 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84189 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1650 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9440 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/63987 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21712 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85153 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1237 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74687 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44268 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1139 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28866 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31540 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29478 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88079 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9327 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6665 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/72368 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9512 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70504 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71587 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17851 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15708 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/14624 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9280 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/14818 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9118 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12647 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10926 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jelee53 commented on Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp:
```diff
@@ -0,0 +1,370 @@
+    m_tokenList = maybeTokenList.releaseReturnValue();
+
+    performParse(context);
+    // FIXME: Figure out a way to provide lvalue of m_result without duplicating the field
```
@szewai do you know how I can do this? 

ExceptionOr requires that I return an lvalue and it's not recommended to WTFMove a class's field.

jelee53 replied:
Also @cdumez  too.

szewai replied:
Does `return m_result;` or `return { m_result  };` work?

jelee53 replied:
For both I still get the same error. 
```
/Users/jessicalee/Documents/webkit-col/Webkit/OpenSource/Source/WebCore/dom/ExceptionOr.h:40:5: note: candidate constructor not viable: no known conversion from 'URLPatternInit' to 'Exception' for 1st argument
   40 |     ExceptionOr(Exception&&);
/Users/jessicalee/Documents/webkit-col/Webkit/OpenSource/Source/WebCore/dom/ExceptionOr.h:41:5: note: candidate constructor not viable: expects an rvalue for 1st argument
   41 |     ExceptionOr(ReturnType&&);
```

jelee53 replied:
[return-m_result.txt](https://github.com/user-attachments/files/18214498/return-m_result.txt)
[return-bracket-m_result.txt](https://github.com/user-attachments/files/18214499/return-bracket-m_result.txt)

szewai replied:
hmm, `return URLPatternInit { m_result };` then

jelee53 replied:
That ended up working! Thank you!

jelee53 commented on Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp:
```diff
@@ -0,0 +1,370 @@
+        }
+
+        if (m_tokenList[m_tokenIndex].type == URLPatternUtilities::TokenType::Open) {
+            m_groupDepth += 1;
```
@cdumez 

Should I change this to `++ m_groupDepth` instead? I recall that you had a preference for pre-increment but I don't remember why. 

In this situation, I actually think it's prettier to do `+= 1` but I don't have a strong preference.

cdumez replied:
we prefer `++m_groupDepth;`

jelee53 replied:
Okay I'll swap out all the `+=` for pre increment `++`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp:
```diff
@@ -0,0 +1,370 @@
+{
+    auto token = getSafeToken(index);
+    // FIXME: Find a better way to convert a char to a string.
+    StringView fixme = StringView(std::span<const char>(&value, 1));
```
maybe:
```
token.value.length() == 1 && token.value[0] == value
```

cdumez replied:
Alternatively, we can update isNonSpecialPatternChar() to take an `ASCIILiteral` instead of a `char`, assert that the length of the ASCIILiteral is 1 and simply use `token.value == value`.

jelee53 replied:
Oh cool I was just about to tag you two! Thanks I'll do the first one ha.

jelee53 commented on Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp:
```diff
@@ -0,0 +1,370 @@
+        }
+        break;
+    case URLPatternConstructorStringParserState::Authority:
+        // Look for identity terminator
```
@szewai is it cool if I leave these comments? 

Basically the reason I had these comments in the first place is because the spec has a bunch of one-liner functions that are helpful for understanding what the function does but create clutter in the code. I wanted to leave these comments to make it clear what I'm doing in this code. (And so if anyone re-visits this code, they'll understand why these functions are "unimplemented"). 

eg:
https://urlpattern.spec.whatwg.org/#is-a-protocol-suffix
https://urlpattern.spec.whatwg.org/#is-an-identity-terminator
<img width="907" alt="Screenshot 2024-12-19 at 11 59 34 AM" src="https://github.com/user-attachments/assets/326839c5-07ae-424d-b0df-842139961e94" />

jelee53 replied:
Also @cdumez  too

szewai replied:
Sure you could leave comments (with period as Chris mentioned).

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp:
```diff
@@ -0,0 +1,370 @@
+// https://urlpattern.spec.whatwg.org/#next-is-authority-slashes
+bool URLPatternConstructorStringParser::isAuthoritySlashesNext() const
+{
+    if (!isNonSpecialPatternChar(m_tokenIndex+1, '/'))
```
Need spaces around `+`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp:
```diff
@@ -0,0 +1,370 @@
+{
+    if (!isNonSpecialPatternChar(m_tokenIndex+1, '/'))
+        return false;
+    if (!isNonSpecialPatternChar(m_tokenIndex+2, '/'))
```
ditto.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp:
```diff
@@ -0,0 +1,370 @@
+    return m_input.substring(componentStartIndex, *token.index - componentStartIndex).toString();
+}
+
+static inline void setInitComponentFromState(URLPatternInit& init, URLPatternConstructorStringParserState state, String componentString)
```
`String` should never be passed by value.

jelee53 replied:
I know this and this was a careless mistake on my part 😓. 

I will change to String&&.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp:
```diff
@@ -0,0 +1,370 @@
+{
+    if (m_state != URLPatternConstructorStringParserState::Init
+        && m_state != URLPatternConstructorStringParserState::Authority
+        && m_state != URLPatternConstructorStringParserState::Done)
```
I *think* coding style says to use curly brackets if the condition is on multiple lines.

jelee53 replied:
You are right! I will change this

>  or a single statement spans multiple lines. 
<img width="931" alt="Screenshot 2024-12-19 at 12 07 45 PM" src="https://github.com/user-attachments/assets/536bac03-7942-49fa-9126-bfbd32e743f1" />

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp:
```diff
@@ -0,0 +1,370 @@
+        // Set init's pathname to empty if conditions are met.
+        static constexpr std::array validStateConditionsForEmptyPathname { URLPatternConstructorStringParserState::Protocol, URLPatternConstructorStringParserState::Authority, URLPatternConstructorStringParserState::Username, URLPatternConstructorStringParserState::Password, URLPatternConstructorStringParserState::Hostname, URLPatternConstructorStringParserState::Port };
+        static constexpr std::array validNewStateConditionsForEmptyPathname { URLPatternConstructorStringParserState::Search, URLPatternConstructorStringParserState::Hash };
+        if (std::find(validStateConditionsForEmptyPathname.begin(), validStateConditionsForEmptyPathname.end(), m_state) != validStateConditionsForEmptyPathname.end()
```
You should consider using `std::ranges::find(validStateConditionsForEmptyPathname, m_state)`. Same below

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp:
```diff
@@ -0,0 +1,370 @@
+}
+
+// https://urlpattern.spec.whatwg.org/#is-a-non-special-pattern-char
+bool URLPatternConstructorStringParser::isNonSpecialPatternChar(size_t index, char value) const
```
`isNonSpecialPatternChararacter`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp:
```diff
@@ -0,0 +1,370 @@
+        }
+        break;
+    case URLPatternConstructorStringParserState::Username:
+        // Look for password prefix
```
Comments in WebKit need to end with a period.

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp:
```diff
@@ -0,0 +1,370 @@
+        break;
+    case URLPatternConstructorStringParserState::Done:
+        // This step is never reached
+        ASSERT_NOT_REACHED();
```
Comments like these are not needed though, since the code is quite self-explanatory.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp:
```diff
@@ -0,0 +1,370 @@
+    if (index < m_tokenList.size())
+        return m_tokenList[index];
+
+    ASSERT(m_tokenList.size() >= 1);
```
Not helpful, I sure `last()` does this internally and in release builds even.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp:
```diff
@@ -0,0 +1,370 @@
+// https://urlpattern.spec.whatwg.org/#make-a-component-string
+String URLPatternConstructorStringParser::makeComponentString() const
+{
+    ASSERT(m_tokenIndex < m_tokenList.size());
```
Not useful, for same reason as above

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp:
```diff
@@ -0,0 +1,370 @@
+    const auto& token = m_tokenList[m_tokenIndex];
+
+    auto componentStartToken = getSafeToken(m_componentStart);
+    ASSERT(componentStartToken.index && token.index);
```
Not useful.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.h:
```diff
@@ -0,0 +1,75 @@
+
+class URLPatternConstructorStringParser {
+public:
+    URLPatternConstructorStringParser(String&& input);
```
`explicit`

cdumez commented on Source/WebCore/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations:
```diff
@@ -47,6 +47,7 @@ Modules/speech/LocalDOMWindowSpeechSynthesis.cpp
 Modules/speech/SpeechSynthesis.cpp
 Modules/streams/TransformStream.cpp
 Modules/webaudio/AudioWorkletGlobalScope.cpp
+Modules/webaudio/BiquadDSPKernel.h
```
No longer needed.

cdumez reviewed (approved):
LGTM with nits.

jelee53 committed:
Implement URLPattern API's Constructor String Parser
https://bugs.webkit.org/show_bug.cgi?id=284911
rdar://141717073

Reviewed by Chris Dumez.

Spec: https://urlpattern.spec.whatwg.org/#constructor-string-parser

* LayoutTests/fast/urlpattern/urlpattern-component-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern-compare.tentative.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern-compare.tentative.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern-compare.tentative.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern-compare.tentative.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern-compare.tentative.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern-compare.tentative.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern-compare.tentative.https.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern-compare.tentative.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.worker-expected.txt:
* Source/WebCore/Modules/url-pattern/URLPattern.cpp:
(WebCore::URLPattern::create):
(WebCore::URLPattern::compileAllComponents):
* Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
* Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp: Added.
(WebCore::URLPatternConstructorStringParser::URLPatternConstructorStringParser):
(WebCore::URLPatternConstructorStringParser::rewind):
(WebCore::URLPatternConstructorStringParser::getSafeToken const):
(WebCore::URLPatternConstructorStringParser::isNonSpecialPatternChararacter const):
(WebCore::URLPatternConstructorStringParser::isSearchPrefix const):
(WebCore::URLPatternConstructorStringParser::isAuthoritySlashesNext const):
(WebCore::URLPatternConstructorStringParser::makeComponentString const):
(WebCore::setInitComponentFromState):
(WebCore::URLPatternConstructorStringParser::computeProtocolMatchSpecialSchemeFlag):
(WebCore::URLPatternConstructorStringParser::changeState):
(WebCore::URLPatternConstructorStringParser::updateState):
(WebCore::URLPatternConstructorStringParser::performParse):
(WebCore::URLPatternConstructorStringParser::parse):
* Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.h: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/288189@main
